[TA3457] feat(istgtcontrol): Add missing metrics

utkarshmani1997 opened the pull request:
On branch US3444-Add-missing-stats
Changes to be committed:
	modified:   data_conn.c
	modified:   istgt_lu.h
	modified:   istgt_lu_ctl.c
	modified:   replication.c
	modified:   replication.h

This commit add the total read/write time stats and uptime into istgtcontrol.
the output is given below.

```
$ sudo istgtcontrol iostats
{ "iqn": "iqn.2016-09.com.openebs.cstor:vol1", "WriteIOPS": "36", "ReadIOPS": "21533",
"TotalWriteBytes": "37441536", "TotalReadBytes": "89096192", "Size": "10737418240",
"UsedLogicalBlocks": "0", "SectorSize": "4096", "UpTime": "48", "TotalReadTime": "72453176",
"TotalWriteTime": "239200860130" }
DONE IOSTATS command
```
Where TotalReadTime and TotalWriteTime is in `nanoseconds`

Signed-off-by: Utkarsh Mani Tripathi <utkarshmani1997@gmail.com>

vishnuitta commented on src/istgt_iscsi.c:
```diff
@@ -5500,6 +5500,9 @@ prof_log(ISTGT_LU_CMD_Ptr p, const char *caller)
 							spec->IO_size[ind].write.tdiff[i].tv_nsec = p->tdiff[i].tv_nsec;
 						}
+						#ifdef REPLICATION
+						spec->totalwritetime += (_r.tv_sec * 1000000000) + (_r.tv_nsec);
```
`_r` is calculated from cmd's `times` array. Don't depend on times array

utkarshmani1997 replied:
ok

vishnuitta reviewed (commented):
time taken to serve an IO shouldn't depend on 'times' data structure. Other changes are fine.

mynktl commented on src/istgt_iscsi.c:
```diff
@@ -5500,6 +5503,11 @@ prof_log(ISTGT_LU_CMD_Ptr p, const char *caller)
 						}
+						#ifdef REPLICATION
+						clock_gettime(CLOCK_MONOTONIC_RAW, &now);
+						diff = (now.tv_sec - start_rw_time.tv_sec)*1000000000 + (now.tv_nsec - start_rw_time.tv_nsec);
```
Better to use MACRO like NANOSEC  instead of a number.
What if `start_rw_time.tv_nsec` is greater than `now.tv_nsec`?
Use `timesdiff` MACRO from `istgt_misc.h`.

mynktl commented on src/istgt_iscsi.c:
```diff
@@ -5500,6 +5503,11 @@ prof_log(ISTGT_LU_CMD_Ptr p, const char *caller)
+						clock_gettime(CLOCK_MONOTONIC_RAW, &now);
+						diff = (now.tv_sec - start_rw_time.tv_sec)*1000000000 + (now.tv_nsec - start_rw_time.tv_nsec);
+						spec->totalwritetime += diff;
+						#endif
```
you are setting the start time for every IO in `istgt_iscsi_read_pdu` and calculating time from `prof_log` which will be called from `sender thread`. This will not give correct information about read/write time.
Better to use an array or new variable in `lu_cmd` to track time per IO.

utkarshmani1997 committed:
[TA3456, TA3457] feat(istgtcontrol): Add missing metrics

This PR add following missing stats
- Total r/w time (nanoseconds)
- Total r/w block count
- Volume uptime (seconds)

output:
```
$ sudo istgtcontrol iostats
{ "iqn": "iqn.2016-09.com.openebs.cstor:vol1", "WriteIOPS": "3931",
  "ReadIOPS": "86", "TotalWriteBytes": "4096000000", "TotalReadBytes": "2146304",
  "Size": "10737418240", "UsedLogicalBlocks": "0", "SectorSize": "4096",
  "UpTime": "91", "TotalReadTime": "84459243", "TotalWriteTime": "560787620941",
  "TotalReadBlockCount": "225443840", "TotalWriteBlockCount": "10304880640"
}
```

Signed-off-by: Utkarsh Mani Tripathi <utkarshmani1997@gmail.com>

vishnuitta reviewed (approved):
changes are good
